### PR TITLE
nanoAOD with proton information built in CMSSW_11_0_0_pre6

### PR DESCRIPTION
**Pull request by jwill24**
#### PR description:

This PR adds the proton information from PPS to the nanoAOD format. This is in response to the issue 332 raised back in May: https://github.com/cms-nanoAOD/cmssw/issues/332

#### PR validation:

This code has been tested on 2017 DoubleEG datasets for both SingleRP and MultiRP proton reconstruction.

### CODE REVIEW

File: PhysicsTools/NanoAOD/plugins/LHCInfoProducer.cc [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+  }
+
+  std::unique_ptr<nanoaod::FlatTable> fillTable(const edm::Event &iEvent, const edm::ESHandle<LHCInfo> & prod) const {
+    
+    const LHCInfo* info = prod.product();
+    float xangle = info->crossingAngle();
+    float betaStar = info->betaStar();
+
+    auto out = std::make_unique<nanoaod::FlatTable>(1,"LHCInfo",true);
+    out->addColumnValue<float>("xangle", xangle, "LHC crossing angle", nanoaod::FlatTable::FloatColumn);
```
Comment (peruzzim) [inline PhysicsTools/NanoAOD/plugins/LHCInfoProducer.cc]: was precision forgotten here?

File: PhysicsTools/NanoAOD/plugins/LHETablesProducer.cc [hunk 2]
```diff
@@ -14,10 +14,10 @@
 class LHETablesProducer : public edm::global::EDProducer<> {
 public:
   LHETablesProducer(edm::ParameterSet const& params)
-      : lheTag_(edm::vector_transform(params.getParameter<std::vector<edm::InputTag>>("lheInfo"),
-                                      [this](const edm::InputTag& tag) { return mayConsume<LHEEventProduct>(tag); })),
-        precision_(params.getParameter<int>("precision")),
-        storeLHEParticles_(params.getParameter<bool>("storeLHEParticles")) {
+    : lheTag_(edm::vector_transform(params.getParameter<std::vector<edm::InputTag>>("lheInfo"),
+				    [this](const edm::InputTag& tag) { return mayConsume<LHEEventProduct>(tag); })),
```
Comment (peruzzim) [inline PhysicsTools/NanoAOD/plugins/LHETablesProducer.cc]: please remove these spurious changes in indentation

File: PhysicsTools/NanoAOD/plugins/ProtonProducer.cc [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+  ProtonProducer( edm::ParameterSet const & ps) :
+    tokenRecoProtons_(consumes<reco::ForwardProtonCollection>(ps.getParameter<edm::InputTag>("tagRecoProtons"))),
+    precision_( ps.getParameter<int>("precision") ),
+    method_( ps.getParameter<std::string>("method") )
+  {
+    produces<edm::ValueMap<int>>("protonRPId");
+    produces<edm::ValueMap<int>>("protonRPType");
+    produces<edm::ValueMap<bool>>("sector45");
+    produces<edm::ValueMap<bool>>("sector56");
+    produces<nanoaod::FlatTable>("trackTable");
```
Comment (peruzzim) [inline PhysicsTools/NanoAOD/plugins/ProtonProducer.cc]: let's use a more proton-specific name for this, please

**Review by peruzzim — CHANGES_REQUESTED** [review]
a first set of comments

**Comment by peruzzim** [comment]
the code should do nothing when no PPS information is found in the event, please use mayConsume and test the validity of the handles when you get data from the event

**Comment by peruzzim** [comment]
there is still something wrong here, @jwill24 can you please take a look?

**Comment by jwill24** [comment]
@peruzzim I checked the failure and it seems that it's an issue of the SimpleFlatTableProducer breaking the nanoSequence when proton information is not available. We discussed this previously and I thought I remember you saying that you could make the necessary changes to PhysicsTools/NanoAOD/interface/SimpleFlatTableProducer.h since I do not have permission to edit this file. It should be as simple as adding an "isValid()" flag.

**Comment by jan-kaspar** [comment]
@peruzzim I am confused by the status of this PR. Are we (PPS) expected to do something now or is your intervention the next step? Thanks in advance for clarification!

**Comment by peruzzim** [comment]
following up on this now - sorry for the delay (no central production was missed because of this)

**Comment by peruzzim** [comment]
My testing fails with a `No "LHCInfoRcd" record found in the EventSetup`

Is this supposed to work also on MC? Does it require specific content in the GT, or customization to the cfg to create this record in the event setup?

**Comment by jan-kaspar** [comment]
> My testing fails with a `No "LHCInfoRcd" record found in the EventSetup`

LHCInfo should be in GT. Alternatively, you can use this tag: LHCInfoEndFill_prompt_v2

> Is this supposed to work also on MC? Does it require specific content in the GT, or customization to the cfg to create this record in the event setup?

PPS has no official simulation yet. Otherwise, from nanoAOD perspective, there should be no difference between data and simulation.

**Comment by peruzzim** [comment]
> > My testing fails with a `No "LHCInfoRcd" record found in the EventSetup`
> 
> LHCInfo should be in GT. Alternatively, you can use this tag: LHCInfoEndFill_prompt_v2
> 

In which GTs should that payload be? In all of them (data and MC)? Was it added recently?

> > Is this supposed to work also on MC? Does it require specific content in the GT, or customization to the cfg to create this record in the event setup?
> 
> PPS has no official simulation yet. Otherwise, from nanoAOD perspective, there should be no difference between data and simulation.

Does this mean that all what is introduced by this PR is supposed to run only on data?

**Comment by jan-kaspar** [comment]
> > > My testing fails with a `No "LHCInfoRcd" record found in the EventSetup`
> > 
> > 
> > LHCInfo should be in GT. Alternatively, you can use this tag: LHCInfoEndFill_prompt_v2
> 
> In which GTs should that payload be? In all of them (data and MC)? Was it added recently?

I have no idea about MC tags.

It should be in data GTs, it was introduced ~year ago. Here's an example I used for some recent proton reco tests:
  https://cms-conddb.cern.ch/cmsDbBrowser/search/Prod/106X_dataRun2_v26

> > > Is this supposed to work also on MC? Does it require specific content in the GT, or customization to the cfg to create this record in the event setup?
> > 
> > 
> > PPS has no official simulation yet. Otherwise, from nanoAOD perspective, there should be no difference between data and simulation.
> 
> Does this mean that all what is introduced by this PR is supposed to run only on data?

At this moment, it can only be useful with data.

**Comment by peruzzim** [comment]
thank you, I will apply some modifications and limit my tests to data

**Comment by jan-kaspar** [comment]
> thank you, I will apply some modifications and limit my tests to data

Thanks to you!

**Comment by peruzzim** [comment]
replaced by #504, with refactored code

**Pull request closed without merging**